go/ql/src/Security/CWE-022/TaintedPathGood.go

```diff
@@ -0,0 +1,20 @@
+package main
+
+import (
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"strings"
+)
+
+func handler(w http.ResponseWriter, r *http.Request) {
+	path := r.URL.Query()["path"][0]
+
+	// GOOD: ensure that the filename has no path separators or parent directory references
+	if strings.Contains(path, "/") || strings.Contains(path, "\\") || strings.Contains(path, "..") {
```

Comment:
Like in the text, it may be good to highlight here that this is only suitable if `path` is expected to refer to a single filename (in case someone looks at this file independently of the `qhelp` context).

Perhaps it would also be good to use the `PathSeparator` constant, instead of hard-coding `/` and `\\`?

Reply:
Perhaps, as a compromise, the check could combine both. Something like:

```go
strings.ContainsAny(path, "/\\") || strings.Contains(path, string(os.PathSeparator)) || strings.Contains(path, "..")
```

or

```go
(dir, filename) := filepath.Split(path); dir != "" || strings.Contains(path, "/") || strings.Contains(path, "..")
```

Reply:
Ah, interesting! Does the query recognise these patterns as sanitisers?

Reply:
It probably _should_, but it looks like neither `ContainsAny` nor performing a check with `Split` is accepted as a sanitiser. For `Contains`, we only seem to accept calls with one of `..`, `../`, or `..\\` as argument.

Reply:
OK, shall we leave the QHelp as it is, then, and tackle the query improvements (and accompanying QHelp rewrites) separately?

Reply:
Sounds OK -- I have created an internal issue to track this.